lighthouse-core/lib/i18n.js

```diff
@@ -42,58 +44,164 @@ const formats = {
+  const valuesForMessageFormat = preprocessMessageValues(template, values);
+
+  const formatter = new MessageFormat(templateForMessageFormat, localeForMessageFormat, formats);
+  const message = formatter.format(valuesForMessageFormat);
+
+  return {message, template: templateForMessageFormat};
+}
+
+/** @param {string[]} pathInLHR */
+function _formatPathAsString(pathInLHR) {
+  let pathAsString = '';
+  for (const property of pathInLHR) {
+    if (/^[a-z]+$/i.test(property)) {
+      if (pathAsString.length) pathAsString += '.';
+      pathAsString += property;
+    } else {
+      if (/]|"|'|\s/.test(property)) throw new Error(`Cannot handle "${property}" in i18n`);
+      pathAsString += `[${property}]`;
+    }
+  }
+
+  return pathAsString;
+}
+
+/**
+ * @return {LH.Locale}
+ */
+function getDefaultLocale() {
+  const defaultLocale = MessageFormat.defaultLocale;
+  if (defaultLocale in LOCALES) return /** @type {LH.Locale} */ (defaultLocale);
+  return 'en-US';
+}
+
+/**
+ * @param {string} filename
+ * @param {Record<string, string>} fileStrings
+ */
+function createStringFormatter(filename, fileStrings) {
+  const mergedStrings = {...UIStrings, ...fileStrings};
+
+  /** @param {string} template @param {*} [values] */
+  const formatFn = (template, values) => {
+    const keyname = Object.keys(mergedStrings).find(key => mergedStrings[key] === template);
+    if (!keyname) throw new Error(`Could not locate: ${template}`);
+
+    const filenameToLookup = keyname in UIStrings ? __filename : filename;
+    const unixStyleFilename = path.relative(LH_ROOT, filenameToLookup).replace(/\\/g, '/');
+    const templateID = `${unixStyleFilename} | ${keyname}`;
+    const templateUsages = formattedStringUsages.get(templateID) || [];
+
+    let indexOfUsage = templateUsages.findIndex(usage => isDeepEqual(usage.values, values));
+    if (indexOfUsage === -1) {
+      templateUsages.push({templateID, template, values});
+      indexOfUsage = templateUsages.length - 1;
+    }
+
+    formattedStringUsages.set(templateID, templateUsages);
+
+    return `${templateID} # ${indexOfUsage}`;
+  };
+
+  return formatFn;
+}
+
+/**
+ * @param {LH.Result} lhr
+ * @param {LH.Locale} locale
+ */
+function replaceLocaleStringReferences(lhr, locale) {
   /**
-   * @param {LH.Locale|null} [newLocale]
+   * @param {*} objectInLHR
+   * @param {LH.I18NMessages} messages
+   * @param {string[]} pathInLHR
    */
-  setLocale(newLocale) {
-    if (!newLocale) return;
-    locale = newLocale;
-  },
+  function replaceInObject(objectInLHR, messages, pathInLHR = []) {
+    if (typeof objectInLHR !== 'object' || !objectInLHR) return;
+
+    for (const [property, value] of Object.entries(objectInLHR)) {
+      const currentPathInLHR = pathInLHR.concat([property]);
+
+      // Check to see if the value in the LHR looks like a string reference. If it is, replace it.
+      if (typeof value === 'string' && /.* \| .* # \d+$/.test(value)) {
```

Comment:
wanna throw in the param groups that you have on 177 so these look more similar?